### PR TITLE
[FAB-18117] Add policy checking to snapshot grpc service

### PR DESCRIPTION
**Pull request by wenjianqiao**
Signed-off-by: Wenjian Qiao wenjianq@gmail.com

#### Type of change
- New feature

#### Description
- Add ACLProvider to snapshot service to check ACL
- Add resources for snapshot submitrequest, cancelrequest and listpending
- Add CheckPolicyNoChannel to ACLProvider interface and implement CheckPolicyNoChannelBySignedData func in policy.go

#### Additional details
This PR is stacked on top of https://github.com/hyperledger/fabric/pull/1790

#### Related issues
https://jira.hyperledger.org/browse/FAB-18117

### CODE REVIEW

File: core/policy/policy.go
Comment (ale-linux) [inline core/policy/policy.go:182]: Does this mean that all the elements in the `signedData` slice must comply with the policy?

File: core/ledger/snapshotgrpc/snapshot_service.go
Comment (ale-linux) [inline core/ledger/snapshotgrpc/snapshot_service.go:115]: Just to double check: this ACL evaluation is not perfomed on the commit path right? Otherwise a time check like this would get us to fork.

File: core/policy/policy.go
Comment (ale-linux) [inline core/policy/policy.go]: This function ends up being called in a ton of places, and this PR changes its failure model (prior to it, we would fail if the supplied channel ID was empty, now we don't). Have you verified that this change does not affect the verifier/committer path? Adding a new function to the 
```
type ACLProvider interface {
	//CheckACL checks the ACL for the resource for the channel using the
	//idinfo. idinfo is an object such as SignedProposal from which an
	//id can be extracted for testing against a policy
	CheckACL(resName string, channelID string, idinfo interface{}) error
}
```
interface (like a `CheckACLNoChannel` function) seems more desirable than changing the contract of the existing function, I'd go with that.

**Review by ale-linux — CHANGES_REQUESTED** [review]

File: core/ledger/snapshotgrpc/snapshot_service.go
Comment (wenjianqiao) [reply core/ledger/snapshotgrpc/snapshot_service.go:116]: @ale-linux Thanks for your comment. I confirm that the events for which the ACL is checked for snapshot service are never on orderer events or on commit path.

File: core/policy/policy.go
Comment (wenjianqiao) [reply core/policy/policy.go:182]: Yes, that's the intention.

File: core/ledger/snapshotgrpc/snapshot_service.go
Comment (wenjianqiao) [reply core/ledger/snapshotgrpc/snapshot_service.go:115]: Same as above.

File: core/ledger/snapshotgrpc/snapshot_service.go
Comment (manish-sethi) [reply core/ledger/snapshotgrpc/snapshot_service.go:116]: I did not notice this expiration check on other APIs that allow admin operations (e.g., join channel). Better not to have a different behavior for snapshot related APIs.

File: core/policy/policy.go
Comment (wenjianqiao) [reply core/policy/policy.go]: When checking existing callers before making the change. I didn't find any code calling CheckACL with []*protoutil.SignedData. The existing code calls CheckACL with either SignedPropoal or Envelope (no []*protoutil.SignedData). However, you have a good point. It is better to add a CheckACLNoChannel and make the explicit call instead of overloading/changing existing behavior.

File: core/ledger/snapshotgrpc/snapshot_service.go
Comment (manish-sethi) [inline core/ledger/snapshotgrpc/snapshot_service.go]: Nit: It's used immediately below, can be used directly in-line

File: core/ledger/snapshotgrpc/snapshot_service_test.go
Comment (manish-sethi) [inline core/ledger/snapshotgrpc/snapshot_service_test.go]: I found this aclprovider error test a bit of a force fit into the existing test case - definition. I feel that having this as independent one makes a better sense.

File: core/policy/policy.go
Comment (manish-sethi) [reply core/policy/policy.go]: Even the existing behavior of the [policy checker](https://github.com/hyperledger/fabric/blob/545754c4bab3311ca56f6098f08dd43cec7c0cf5/core/policy/policy.go#L38) is inconsistent. Two similar functions [CheckPolicy](https://github.com/hyperledger/fabric/blob/545754c4bab3311ca56f6098f08dd43cec7c0cf5/core/policy/policy.go#L51) and [CheckPolicyBySignedData](https://github.com/hyperledger/fabric/blob/545754c4bab3311ca56f6098f08dd43cec7c0cf5/core/policy/policy.go#L145) behave differently, if no channelID is specified. 

@ale-linux  - I agree that adding the exposing the function `CheckACLNoChannel` at the `ACLProvider` level makes it clearer but I would add a note (I assume that you meant this implicitly as well) that we should modify the existing callers go via this function instead of first function above. Otherwise, it will simply add to the confusion looking at the code for - which would look like
- if you work with `SignedProposal` and have no channel call the function `CheckACL` with channelID as empty string
- if you work with `SignedData`, use the function `CheckACLNoChannel`.
This modification can be done in a separate PR and can open a Jira for this.

File: core/ledger/snapshotgrpc/snapshot_service.go
Comment (ale-linux) [reply core/ledger/snapshotgrpc/snapshot_service.go:116]: +1 @manish-sethi, good point

File: core/policy/policy.go
Comment (ale-linux) [reply core/policy/policy.go]: Yes @manish-sethi, I share your view and the approach you propose

Comment (wenjianqiao) [reply core/policy/policy.go]: @manish-sethi agree, I have thought the same and will open a Jira for it.

File: core/ledger/snapshotgrpc/snapshot_service.go
Comment (wenjianqiao) [reply core/ledger/snapshotgrpc/snapshot_service.go:112]: @manish-sethi The other APIs that allow admin ops are done via endorsor ProcessProposal that checks expiration via an auth filter [expiration.go](https://github.com/hyperledger/fabric/blob/master/core/handlers/auth/filter/expiration.go#L49) Therefore, individual cmd doesn't need separate call to check certification expiration. However, because snapshot service is a separate grpc service, we should add this check.

Comment (wenjianqiao) [reply core/ledger/snapshotgrpc/snapshot_service.go:116]: @manish-sethi The other APIs that allow admin ops are done via endorsor ProcessProposal that checks expiration via an auth filter expiration.go Therefore, individual cmd doesn't need separate call to check certificate expiration. However, because snapshot service is a separate grpc service, we should add this check.

Comment (manish-sethi) [reply core/ledger/snapshotgrpc/snapshot_service.go:116]: I see, I missed that part here. Thanks for the clarification @wenjianqiao

Comment (manish-sethi) [reply core/ledger/snapshotgrpc/snapshot_service.go:116]: Just to take a note, ideally, that should also have been performed at one place like ACLCheck but not related to this PR.

**Comment by wenjianqiao** [comment]
@ale-linux @manish-sethi I have addressed your comments. Also opened https://jira.hyperledger.org/browse/FAB-18185 to update the existing code to call CheckACLNoChannel when there is no channel.

**Review by ale-linux — COMMENTED** [review]
LGTM. I won't merge yet to give @manish-sethi a chance to review as well.

**Review by ale-linux — DISMISSED** [review]

File: core/policy/policy.go
Comment (manish-sethi) [reply core/policy/policy.go]: missed rolling back this change?

Comment (manish-sethi) [inline core/policy/policy.go]: missed rolling back this change?

Comment (wenjianqiao) [reply core/policy/policy.go]: Good catch. Done.

File: core/aclmgmt/resourceprovider.go
Comment (manish-sethi) [inline core/aclmgmt/resourceprovider.go]: This really looks bad... an implementation knows the behavior of the consumer (which is usually the other way around). I am not well versed with this package but is it not possible for this impl to implement a separate interface which does not have this function - or rather does it have to implement this interface at all? I see this getting used only at one place in the production code where it does not need to implement this interface.

Comment (wenjianqiao) [reply core/aclmgmt/resourceprovider.go]: Removed because peer resource does not allow override and never gets here.

File: core/policy/policy_test.go
Comment (manish-sethi) [inline core/policy/policy_test.go:200]: Nit: To differentiate, better to use different names, perhaps `CheckPolicyNoChannelSignedProposal` and `CheckPolicyNoChannelSignedData`. But this should not hold this PR from merging.

**Review by manish-sethi — APPROVED** [review]

**Pull request merged**
